File: tests/agents/memory/TestDeque.py
import pytest

from benchmarks.agents.memory.Deque import Deque
from collections import deque


class TestDeque:
    """
    Class testing the double ended queue.
    """

    @staticmethod
    def compare(queue_1, queue_2):
        """
        Compare the queues passed as parameters.
        :param queue_1: the first queue
        :param queue_2: the second queue
        """
        assert len(queue_1) == len(queue_2)
        for i in range(len(queue_1)):
            assert queue_1[i] == queue_2[i]

    @pytest.mark.parametrize("elements, max_size", [
        ([5, 1, 1], 2), ([], 4), ([1, 2, 3, 4], 4), ([10, 2], 4),
    ])
    def test_clear(self, elements, max_size):

        # Create two deque (one from my c++ library and one from python standard library).
        queue = Deque(max_size=max_size)
        python_queue = deque(maxlen=max_size)

        # Add all elements to the end of the double ended queues.
        for element in elements:
            queue.append(element)
            python_queue.append(element)

        # Clear both double ended queues.
        queue.clear()
        python_queue.clear()

        # Check that both queues are identical.
        self.compare(queue, python_queue)

    @pytest.mark.parametrize("elements, max_size", [
        ([5, 1, 1], 2), ([], 4), ([1, 2, 3, 4], 4), ([10, 2], 4),
    ])
    def test_append(self, elements, max_size):

        # Create two deque (one from my c++ library and one from python standard library).
        queue = Deque(max_size=max_size)
        python_queue = deque(maxlen=max_size)

        # Add all elements to the end of the double ended queues.
        for element in elements:
            queue.append(element)
            python_queue.append(element)

        # Check that both queues are identical.
        self.compare(queue, python_queue)

    @pytest.mark.parametrize("elements, max_size", [
        ([5, 1, 1], 2), ([], 4), ([1, 2, 3, 4], 4), ([10, 2], 4),
    ])
    def test_append_left(self, elements, max_size):

        # Create two deque (one from my c++ library and one from python standard library).
        queue = Deque(max_size=max_size)
        python_queue = deque(maxlen=max_size)

        # Add all elements to the end of the double ended queues.
        for element in elements:
            queue.append_left(element)
            python_queue.appendleft(element)

        # Check that both queues are identical.
        self.compare(queue, python_queue)

    @pytest.mark.parametrize("elements, n_pops, max_size", [
        ([5, 1, 1], 1, 2), ([1, 2, 3, 4], 3, 4), ([10, 2], 2, 4),
    ])
    def test_pop_left(self, elements, n_pops, max_size):

        # Create two deque (one from my c++ library and one from python standard library).
        queue = Deque(max_size=max_size)
        python_queue = deque(maxlen=max_size)

        # Add all elements to the end of the double ended queues.
        for element in elements:
            queue.append_left(element)
            python_queue.appendleft(element)

        # Remove elements from the front of the double ended queues.
        for _ in range(n_pops):
            queue.pop_left()
            python_queue.popleft()

        # Check that both queues are identical.
        self.compare(queue, python_queue)

    @pytest.mark.parametrize("elements, n_pops, max_size", [
        ([5, 1, 1], 1, 2), ([1, 2, 3, 4], 3, 4), ([10, 2], 2, 4),
    ])
    def test_pop(self, elements, n_pops, max_size):

        # Create two deque (one from my c++ library and one from python standard library).
        queue = Deque(max_size=max_size)
        python_queue = deque(maxlen=max_size)

        # Add all elements to the end of the double ended queues.
        for element in elements:
            queue.append_left(element)
            python_queue.appendleft(element)

        # Remove elements from the front of the double ended queues.
        for _ in range(n_pops):
            queue.pop()
            python_queue.pop()

        # Check that both queues are identical.
        self.compare(queue, python_queue)

    @pytest.mark.parametrize("elements, max_size, length", [
        ([5, 1, 1], 2, 2), ([1, 2, 3, 4], 4, 4), ([10, 2], 4, 2),
    ])
    def test_len(self, elements, max_size, length):

        # Create two deque (one from my c++ library and one from python standard library).
        queue = Deque(max_size=max_size)
        python_queue = deque(maxlen=max_size)

        # Add all elements to the end of the double ended queues.
        for element in elements:
            queue.append(element)
            python_queue.append(element)

        # Check that both queues have the expected length.
        assert len(queue) == len(python_queue) == length

File: tests/src/agents/memory/test_deque.cpp
#include <gtest/gtest.h>
#include <torch/extension.h>
#include "agents/memory/test_deque.hpp"
#include "helpers/deque.hpp"
#include "relab_test.hpp"

namespace relab::test::agents::memory {

    DequeParameters::DequeParameters(const std::initializer_list<int> &elements, int max_size, int n_pops, int length)
        : elements(elements), max_size(max_size), n_pops(n_pops), length(length) {}

    DequeParameters::DequeParameters() : DequeParameters({}, 0, 0, 0) {}

    void TestDeque::compare(Deque<int> &queue, const std::deque<int> &result) {
        EXPECT_EQ(queue.size(), result.size());
        for (size_t i = 0; i < queue.size(); i++) {
            EXPECT_EQ(queue[i], result[i]);
        }
    }

    std::deque<int> TestDeque::getResult(const DequeParameters &params, const PushType &push_type, const PopType &pop_type) {

        // Create the result deque.
        std::deque<int> result;

        // Add all elements to the result deque.
        for (auto element : params.elements) {
            if (push_type == PushType::FRONT) {
                result.push_front(element);
            } else {
                result.push_back(element);
            }
        }

        // Remove elements from the result deque to ensure the queue size is inferior to the max size.
        while (static_cast<int>(result.size()) > params.max_size) {
            if (push_type == PushType::FRONT) {
                result.pop_back();
            } else {
                result.pop_front();
            }
        }

        // Remove elements from the result deque.
        if (pop_type == PopType::NO_POP) {
            return result;
        }
        for (int i = 0; i < params.n_pops; i++) {
            if (pop_type == PopType::FRONT) {
                result.pop_front();
            } else {
                result.pop_back();
            }
        }
        return result;
    }

    void TestDeque::SetUp() {

        // Create a deque.
        this->params = GetParam();
        this->queue = std::make_unique<Deque<int>>(params.max_size);
    }

    TEST_P(TestDeque, TestClear) {

        // Add all elements to the end of the double ended queue.
        for (auto element : params.elements) {
            queue->push_back(element);
        }

        // Clear the double ended queue.
        queue->clear();

        // Check that the queue matches the expected result.
        std::deque<int> result;
        TestDeque::compare(*queue, result);
    }

    TEST_P(TestDeque, TestPushBack) {

        // Add all elements to the end of the double ended queue.
        for (auto element : params.elements) {
            queue->push_back(element);
        }

        // Check that the queue matches the expected result.
        auto result = TestDeque::getResult(params, PushType::BACK, PopType::NO_POP);
        TestDeque::compare(*queue, result);
    }

    TEST_P(TestDeque, TestPushFront) {

        // Add all elements to the front of the double ended queue.
        for (auto element : params.elements) {
            queue->push_front(element);
        }

        // Check that the queue matches the expected result.
        auto result = TestDeque::getResult(params, PushType::FRONT, PopType::NO_POP);
        TestDeque::compare(*queue, result);
    }

    TEST_P(TestDeque, TestPopFront) {

        // Add all elements to the end of the double ended queue.
        for (auto element : params.elements) {
            queue->push_front(element);
        }

        // Remove elements from the front of the double ended queue.
        for (int i = 0; i < params.n_pops; i++) {
            queue->pop_front();
        }

        // Check that the queue matches the expected result.
        auto result = TestDeque::getResult(params, PushType::FRONT, PopType::FRONT);
        TestDeque::compare(*queue, result);
    }

    TEST_P(TestDeque, TestPopBack) {

        // Add all elements to the end of the double ended queue.
        for (auto element : params.elements) {
            queue->push_front(element);
        }

        // Remove elements from the front of the double ended queue.
        for (int i = 0; i < params.n_pops; i++) {
            queue->pop_back();
        }

        // Check that the queue matches the expected result.
        auto result = TestDeque::getResult(params, PushType::FRONT, PopType::BACK);
        TestDeque::compare(*queue, result);
    }

    TEST_P(TestDeque, TestSize) {

        // Add all elements to the end of the double ended queue.
        for (auto element : params.elements) {
            queue->push_back(element);
        }

        // Check that the queue's size is correct.
        EXPECT_EQ(queue->size(), params.length);
    }

    INSTANTIATE_TEST_SUITE_P(UnitTests, TestDeque, testing::Values(
        DequeParameters({5, 1, 1}, 2, 1, 2), DequeParameters({}, 4, 0, 0),
        DequeParameters({1, 2, 3, 4}, 4, 3, 4), DequeParameters({10, 2}, 4, 2, 2)
    ));
}
